File: ios/RecipeClipper/Data/Remote/RedditRecipeParser.swift
import Foundation

/// A Reddit post's `.json` listing to a recipe. Ported from Android's `RedditRecipeParser`.
/// The response is a two-element array, the post (a `t3`) then its comment tree (`t1`s with
/// their `replies`). In order:
///  1. **The post body**, if `selftext` splits cleanly (`RecipeTextSplitter`).
///  2. **A transcription in the comments**: the best comment that splits (`RedditCommentScorer`).
///  3. **Nothing found**: `.noTranscription`, with the post's title and photo. A legitimate
///     outcome, not a failure: never guess a recipe from prose.
/// The name is the post title, the photo the post's image; a crosspost with no body or photo
/// of its own borrows the original's. Anything that isn't a post listing is `.noRecipeFound`.
///
/// Pure: JSON text in, a `ParseResult` out.
enum RedditRecipeParser {

    /// How deep the comment walk goes. Reddit itself stops nesting long before this.
    private static let maxDepth = 50

    static func parse(_ json: String, sourceUrl: String) -> ParseResult {
        // The JSON-LD parser's reader: it refuses absurd nesting before JSONSerialization runs.
        guard let listings = JsonLdRecipeParser.parseJson(json) as? [Any],
              let first = listings.first as? [String: Any],
              let child = ((first["data"] as? [String: Any])?["children"] as? [Any])?.first as? [String: Any],
              str(child, "kind") == "t3",
              let post = child["data"] as? [String: Any]
        else { return .error(.noRecipeFound) }

        let title = JsonLdRecipeParser.stripHtml(str(post, "title"))
        if title.kIsBlank { return .error(.noRecipeFound) }
        let original = (post["crosspost_parent_list"] as? [Any])?.first as? [String: Any]
        let image = imageOf(post) ?? original.flatMap(imageOf)

        var body = bodyOf(post)
        if body.kIsBlank { body = original.map(bodyOf) ?? "" }
        let commentListing = listings.count > 1 ? listings[1] as? [String: Any] : nil
        guard let split = RecipeTextSplitter.split(body)
                ?? RedditCommentScorer.pick(comments(commentListing)).flatMap(RecipeTextSplitter.split)
        else { return .error(.noTranscription(title: title, imageUrl: image)) }

        return .success(Recipe(
            name: title,
            image: image,
            ingredients: split.ingredients,
            instructions: split.instructions,
            prepTime: split.prepTime,
            cookTime: split.cookTime,
            totalTime: split.totalTime,
            yield: split.yield,
            sourceUrl: sourceUrl,
            sourceType: .reddit
        ))
    }

    /// A string field, or "" when it's absent, null or not a string.
    private static func str(_ json: [String: Any], _ key: String) -> String {
        json[key] as? String ?? ""
    }

    private static func bodyOf(_ post: [String: Any]) -> String {
        let text = str(post, "selftext")
        let trimmed = text.kTrimmed
        return trimmed == "[removed]" || trimmed == "[deleted]" ? "" : text
    }

    /// Every comment body in the tree, depth first, in the order Reddit sent them.
    static func comments(_ listing: [String: Any]?) -> [String] {
        var out: [String] = []
        func walk(_ node: [String: Any]?, _ depth: Int) {
            guard let node, depth <= maxDepth,
                  let children = (node["data"] as? [String: Any])?["children"] as? [Any] else { return }
            for case let child as [String: Any] in children {
                guard str(child, "kind") == "t1", let data = child["data"] as? [String: Any] else { continue }
                let body = str(data, "body")
                if !body.kIsBlank { out.append(body) }
                walk(data["replies"] as? [String: Any], depth + 1) // "" when there are none
            }
        }
        walk(listing, 0)
        return out
    }

    private static let imageExtension = JRegex(#"\.(?:jpe?g|png|gif|webp)$"#, ignoreCase: true)

    /// The post's photo: Reddit's preview of it, else a gallery's first picture, else a link
    /// straight to an image. `&amp;` in these URLs is undone, and nothing else: it's a URL.
    static func imageOf(_ post: [String: Any]) -> String? {
        if let images = (post["preview"] as? [String: Any])?["images"] as? [Any],
           let source = (images.first as? [String: Any])?["source"] as? [String: Any] {
            let url = str(source, "url")
            if url.hasPrefix("http") { return unescape(url) }
        }

        if let items = (post["gallery_data"] as? [String: Any])?["items"] as? [Any],
           let firstId = (items.first as? [String: Any]).map({ str($0, "media_id") }), !firstId.isEmpty,
           let s = ((post["media_metadata"] as? [String: Any])?[firstId] as? [String: Any])?["s"] as? [String: Any] {
            let u = str(s, "u")
            let candidate = u.isEmpty ? str(s, "gif") : u
            if candidate.hasPrefix("http") { return unescape(candidate) }
        }

        let overridden = str(post, "url_overridden_by_dest")
        let link = overridden.isEmpty ? str(post, "url") : overridden
        let path = URLComponents(string: link)?.path ?? ""
        return link.hasPrefix("http") && imageExtension.containsMatch(in: path) ? unescape(link) : nil
    }

    private static func unescape(_ url: String) -> String {
        url.replacingOccurrences(of: "&amp;", with: "&")
    }
}

File: app/src/main/java/com/example/recipeclipper/data/remote/RedditRecipeParser.kt
package com.example.recipeclipper.data.remote

import com.example.recipeclipper.data.model.ParseError
import com.example.recipeclipper.data.model.ParseResult
import com.example.recipeclipper.data.model.Recipe
import com.example.recipeclipper.data.model.SourceType
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.json.JSONTokener
import org.jsoup.Jsoup

/**
 * A Reddit post's `.json` listing to a recipe. The response is a two-element array: the post
 * (a `t3`), then its comment tree (`t1`s, each with its `replies`). In order:
 *
 *  1. **The post body.** If `selftext` splits cleanly ([RecipeTextSplitter]), that's the recipe.
 *  2. **A transcription in the comments.** Otherwise every comment in the tree is scored and the
 *     best one that splits wins ([RedditCommentScorer]).
 *  3. **Nothing found:** [ParseError.NoTranscription], with the post's title and photo. A
 *     legitimate outcome, not a failure: never guess a recipe from prose.
 *
 * The recipe's name is the post title; its photo is the post's image. A crosspost with no body
 * or photo of its own borrows the original's. Text that isn't a post listing at all is
 * [ParseError.NoRecipeFound].
 *
 * Pure: JSON text in, a [ParseResult] out.
 */
object RedditRecipeParser {

    /** How deep the comment walk goes. Reddit itself stops nesting long before this. */
    private const val MAX_DEPTH = 50

    fun parse(json: String, sourceUrl: String): ParseResult {
        val root = try {
            JSONTokener(json).nextValue()
        } catch (e: JSONException) {
            null
        } catch (e: StackOverflowError) {
            // org.json recurses while parsing, so absurd nesting overflows before any guard of
            // ours runs. Narrower than JsonLdRecipeParser's Throwable, with the same reasoning.
            null
        }
        val listings = root as? JSONArray ?: return ParseResult.Error(ParseError.NoRecipeFound)
        val post = listings.optJSONObject(0)
            ?.optJSONObject("data")?.optJSONArray("children")?.optJSONObject(0)
            ?.takeIf { it.optString("kind") == "t3" }?.optJSONObject("data")
            ?: return ParseResult.Error(ParseError.NoRecipeFound)

        val title = stripHtml(post.optString("title")).ifBlank {
            return ParseResult.Error(ParseError.NoRecipeFound)
        }
        val original = post.optJSONArray("crosspost_parent_list")?.optJSONObject(0)
        val image = imageOf(post) ?: original?.let(::imageOf)

        val body = bodyOf(post).ifBlank { original?.let(::bodyOf).orEmpty() }
        val split = RecipeTextSplitter.split(body)
            ?: RedditCommentScorer.pick(comments(listings.optJSONObject(1)))?.let(RecipeTextSplitter::split)
            ?: return ParseResult.Error(ParseError.NoTranscription(title, image))

        return ParseResult.Success(
            Recipe(
                name = title,
                image = image,
                ingredients = split.ingredients,
                instructions = split.instructions,
                prepTime = split.prepTime,
                cookTime = split.cookTime,
                totalTime = split.totalTime,
                yield = split.yield,
                sourceUrl = sourceUrl,
                sourceType = SourceType.REDDIT
            )
        )
    }

    private fun stripHtml(raw: String): String = Jsoup.parse(raw).text()

    private fun bodyOf(post: JSONObject): String =
        post.optString("selftext").takeUnless { it.trim() == "[removed]" || it.trim() == "[deleted]" }.orEmpty()

    /** Every comment body in the tree, depth first, in the order Reddit sent them. */
    internal fun comments(listing: JSONObject?): List<String> {
        val out = mutableListOf<String>()
        fun walk(node: JSONObject?, depth: Int) {
            if (node == null || depth > MAX_DEPTH) return
            val children = node.optJSONObject("data")?.optJSONArray("children") ?: return
            for (i in 0 until children.length()) {
                val child = children.optJSONObject(i) ?: continue
                if (child.optString("kind") != "t1") continue // "more" stubs have no text
                val data = child.optJSONObject("data") ?: continue
                val body = data.optString("body")
                if (body.isNotBlank()) out.add(body)
                walk(data.optJSONObject("replies"), depth + 1) // "" when there are none
            }
        }
        walk(listing, 0)
        return out
    }

    private val IMAGE_EXTENSION = Regex("\\.(?:jpe?g|png|gif|webp)$", RegexOption.IGNORE_CASE)

    /**
     * The post's photo: the preview Reddit renders for it, else the first picture of a
     * gallery, else a link straight to an image. Reddit escapes `&` in these URLs even with
     * `raw_json` off, so `&amp;` is undone (and nothing else is touched: it's a URL).
     */
    internal fun imageOf(post: JSONObject): String? {
        post.optJSONObject("preview")?.optJSONArray("images")?.optJSONObject(0)
            ?.optJSONObject("source")?.optString("url")?.takeIf { it.startsWith("http") }
            ?.let { return unescape(it) }

        val firstId = post.optJSONObject("gallery_data")?.optJSONArray("items")
            ?.optJSONObject(0)?.optString("media_id")
        if (!firstId.isNullOrEmpty()) {
            val s = post.optJSONObject("media_metadata")?.optJSONObject(firstId)?.optJSONObject("s")
            (s?.optString("u")?.ifEmpty { null } ?: s?.optString("gif")?.ifEmpty { null })
                ?.takeIf { it.startsWith("http") }
                ?.let { return unescape(it) }
        }

        val link = post.optString("url_overridden_by_dest").ifEmpty { post.optString("url") }
        val path = try { java.net.URI(link).path.orEmpty() } catch (e: Exception) { "" }
        return link.takeIf { it.startsWith("http") && IMAGE_EXTENSION.containsMatchIn(path) }?.let(::unescape)
    }

    private fun unescape(url: String) = url.replace("&amp;", "&")
}
